### PR TITLE
feat: jwt_backends - create backend mechanism and add authlib support

### PR DESCRIPTION
**Pull request by hasB4K**
Hello @k4black,

I've implemented the changes discussed in issue #40. The idea is to add a JWTBackend concept to fastapi-jwt for better security and customization. This allows using other authentication libraries besides python-jose, addressing the security concerns mentioned in #40.

Here's what's new in this PR:

* A `jwt_backends` folder with `PythonJoseJWTBackend` and `AuthlibJWTBackend`.
* A dynamic import system in `jwt_backends/__init__.py`.
* A `define_default_jwt_backend` function, so users can pick their preferred backend. I think most will either stick with the default or set it once.
* I've kept `python-jose` support for now, given its popularity, but we might want to consider a deprecation warning.

I think this approach could also help with:
* #7: making it easier to create custom backends.
* #26: allowing decode_kwargs customization and the "subject" to "sub" swap. Though, standardizing on "sub" might be simpler, that's a side point.

I hope this PR helps.

Have a great week!
Kind regards,

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -37,7 +36,15 @@ documentation = "https://k4black.github.io/fastapi-jwt/"
 
 
 [project.optional-dependencies]
+authlib = [
+    "Authlib >=1.3.0"
+]
+python_jose = [
+    "python-jose[cryptography] >=3.3.0"
+]
```
Comment (hasB4K) [inline pyproject.toml:44]: useful for people to be able to do:
pip install fastapi_jwt[authlib]

or 

pip install fastapi_jwt[python_jose]

File: fastapi_jwt/jwt.py [hunk 2]
```diff
@@ -8,11 +8,21 @@
...
+
+def define_default_jwt_backend(cls):
+    global DEFAULT_JWT_BACKEND
+    DEFAULT_JWT_BACKEND = cls
+
+
+if AuthlibJWTBackend is not None:
+    define_default_jwt_backend(AuthlibJWTBackend)
+elif PythonJoseJWTBackend is not None:
+    define_default_jwt_backend(PythonJoseJWTBackend)
```
Comment (hasB4K) [inline fastapi_jwt/jwt.py:25]: Authlib is the default if both authlib and python-jose are installed.

File: tests/test_security_jwt_general_optional.py [hunk 3]
```diff
@@ -196,27 +188,29 @@ def test_security_jwt_access_token_changed():
...
+    mock_now_for_backend(mocker, jwt_backend, minutes=3)  # 3 min left
     response = client.get(
         "/users/me", headers={"Authorization": f"Bearer {access_token}"}
     )
     assert response.status_code == 200, response.text
     assert response.json() == {"username": "username", "role": "user"}
 
-    mocker.patch("jose.jwt.datetime", _FakeDateTimeLong)  # 42 days left
-
+    mock_now_for_backend(mocker, jwt_backend, days=42)  # 42 days left
```
Comment (hasB4K) [inline tests/test_security_jwt_general_optional.py:203]: all the tests changes are about:
* make it parametrize to test the two backends
* mock datetime.now or time.time depending of the backend used (if necessary)
* create a client depending of the backend used (need to be dynamic and wrapped inside a `create_example_client` to really test the backends)

File: tests/test_security_jwt_general_optional.py [hunk 4]
```diff
@@ -266,19 +266,22 @@ def test_security_jwt_refresh_token_changed():
...
         "/refresh", headers={"Authorization": f"Bearer {refresh_token}"}
     )
     assert response.status_code == 200, response.text
     assert response.json() == {"msg": "Create an account first"}
 
 
-def test_security_jwt_custom_jti():
+@pytest.mark.parametrize("jwt_backend", [AuthlibJWTBackend, PythonJoseJWTBackend])
+def test_security_jwt_custom_jti(jwt_backend):
+    client, unique_identifiers_database = create_example_client(jwt_backend)
```
Comment (hasB4K) [inline tests/test_security_jwt_general_optional.py:284]: and for two tests we need to expose unique_identifiers_database with the client

**Comment by rjjanuary** [comment]
I'm looking at moving to JWT's for a few FastAPI projects and was considering building my own framework.  Looking through fastapi-jwt, I love the design, however also feel that it's not in our best interest to rely on python-jose. Has there been any consideration of merging this into the project?  If so, is there any additional work needed I could assist with?

**Comment by hasB4K** [comment]
@rjjanuary, it seems that the maintainer of this project is not very active.
Regarding this PR, I don't think it will be merged, and tbh I would be in favor of a full deprecation of `python-jose` entirely.

I'm now using [fastapi-users](https://github.com/fastapi-users/fastapi-users) which provide an auth framework in which you can select what you want and what you want to change. The project is actively maintained too. They had a similar issue with `passlib` being unmaintained, and the maintainer of fastapi-users handled the situation very well. See: https://github.com/fastapi-users/fastapi-users/issues/1325

**Comment by rjjanuary** [comment]
Thank you @hasB4K,  I appreciate the quick response.  I'll look deeper into fastapi-users, however at first glance it may be a little overkill for current projects.

**Comment by k4black** [comment]
@hasB4K Thank you for the great contribution!

I'll revise design of backends a bit to use less =None hacks (Keep backends through error if lib not installed and make jwt.py itself select backend) and push to this branch today!

**Comment by k4black** [comment]
@hasB4K Thank you for the great contribution!
Sorry for not being really active, if you eager to use and update this project i can grant you rights. Please write me email at kdchernyshev at gmail dot com and we can have a short call.

Regarding your contribution:
I'll update backends structure a bit to make jwt.py make a backend choice. For now i'll keep backends feature, but in general I agree with deprecation and removal of jose.

**Pull request merged**